python/dgl/graphbolt/subgraph_sampler.py

```diff
@@ -191,6 +210,19 @@ def _node_pairs_preprocess(self, minibatch):
             compacted_negative_dsts if has_neg_dst else None,
         )
 
+    def _sample(self, minibatch):
+        (
+            minibatch.input_nodes,
+            minibatch.sampled_subgraphs,
+        ) = self.sample_subgraphs(
+            minibatch._seed_nodes, minibatch._seeds_timestamp
+        )
+        return minibatch
+
+    def sampling_stages(self, datapipe):
+        """The sampling stages are defined here by chaining to the datapipe."""
+        return datapipe.transform(self._sample)
+
     def sample_subgraphs(self, seeds, seeds_timestamp):
```

Comment:
is this API still required or the recommended way to customize `SubgraphSampelr`? Even for the canonical ones we offers(`NeighborSampler`) don't override it.

Reply:
my concern is the canonical neighbor samplers does not follow it though overriding `sample_subgraphs()` in customized sampler could works as expected.

Reply:
If the user wants to take advantage of the optimizations, they need to override sampling_stages and use the appropriate building blocks. Otherwise, overriding sample_subgraphs is completely fine. Maybe, if we can refine the sampling_stages building blocks even further, we can remove sample_subgraphs and ask the user to always make use of the building blocks.

Reply:
We can think about this further and refine SubgraphSampler even further before the release.

Reply:
We will probably turn SamplePerLayer into an Abstract class that the users can derive from. Then, we can add a version of it converting into a dgl.sparse datastructure for customized sampling algorithm.

Reply:
Maybe then we can make `sample_subgraphs` into an internal method. I will open further PRs to explore these ideas.

Reply:
the question is more about what APIs we offer in the base SubgraphSampler for customization. I'm ok with current implementation though the co-existence of `sample_subgraphs()` and `sampling_stages()` seems not a final implementation.

Reply:
I see. Let's make further progress on these files and let's see how the final implementation looks like. When we have everything we wanted to have, we can take a better look and make a better final decision about the API before the release.